fpa/service.go

```diff
@@ -0,0 +1,164 @@
+
+func (s *Service) Info(ctx context.Context, req *pb.InfoRequest) (*pb.InfoReply, error) {
+	i, err := s.getInstanceByToken(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	info, err := i.Info(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	reply := &pb.InfoReply{
+		Id:   info.ID.String(),
+		Pins: make([]string, len(info.Pins)),
+		Wallet: &pb.WalletInfo{
+			Address: info.Wallet.Address,
+			Balance: info.Wallet.Balance,
+		},
+	}
+	for i, p := range info.Pins {
+		reply.Pins[i] = p.String()
+	}
+
+	return reply, nil
+}
+
+func (s *Service) Show(ctx context.Context, req *pb.ShowRequest) (*pb.ShowReply, error) {
+	i, err := s.getInstanceByToken(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	c, err := cid.Decode(req.GetCid())
+	if err != nil {
+		return nil, err
+	}
+
+	info, err := i.Show(c)
+	if err != nil {
+		return nil, err
+	}
+	reply := &pb.ShowReply{
+		Cid:     info.Cid.String(),
+		Created: info.Created.UnixNano(),
+		Hot: &pb.ShowReply_HotInfo{
+			Size: int64(info.Hot.Size),
+			Ipfs: &pb.ShowReply_IpfsHotInfo{
+				Created: info.Hot.Ipfs.Created.UnixNano(),
+			},
+		},
+		Cold: &pb.ShowReply_ColdInfo{
+			Filecoin: &pb.ShowReply_FilInfo{
+				PayloadCid: info.Cold.Filecoin.PayloadCID.String(),
+				Duration:   info.Cold.Filecoin.Duration,
+				Proposals:  make([]*pb.ShowReply_FilStorage, len(info.Cold.Filecoin.Proposals)),
+			},
+		},
+	}
+	for i, p := range info.Cold.Filecoin.Proposals {
+		reply.Cold.Filecoin.Proposals[i] = &pb.ShowReply_FilStorage{
+			ProposalCid: p.ProposalCid.String(),
+			Failed:      p.Failed,
+		}
+	}
+
+	return reply, nil
+}
+
+func (s *Service) Store(ctx context.Context, req *pb.StoreRequest) (*pb.StoreReply, error) {
+	i, err := s.getInstanceByToken(ctx)
+	if err != nil {
+		return nil, err
+	}
+	c, err := cid.Decode(req.GetCid())
+	if err != nil {
+		return nil, err
+	}
+	if err := i.Put(ctx, c); err != nil {
+		return nil, err
+	}
+	return &pb.StoreReply{}, nil
+}
+
+func (s *Service) Get(req *pb.GetRequest, srv pb.API_GetServer) error {
+	i, err := s.getInstanceByToken(srv.Context())
+	if err != nil {
+		return err
+	}
+	c, err := cid.Decode(req.GetCid())
+	if err != nil {
+		return err
+	}
+	r, err := i.Get(srv.Context(), c)
+	if err != nil {
+		return err
+	}
+
+	buffer := make([]byte, 1024*32)
+	for {
+		bytesRead, err := r.Read(buffer)
+		if err != nil && err != io.EOF {
+			return err
+		}
+		if sendErr := srv.Send(&pb.GetReply{Chunk: buffer[:bytesRead]}); sendErr != nil {
+			return sendErr
+		}
+		if err == io.EOF {
+			return nil
+		}
+	}
+}
+
+func (s *Service) Create(ctx context.Context, req *pb.CreateRequest) (*pb.CreateReply, error) {
+	id, addr, err := s.m.Create(ctx)
+	if err != nil {
+		log.Errorf("creating instance: %s", err)
+		return nil, err
+	}
+	return &pb.CreateReply{
+		Id:      id.String(),
+		Address: addr,
+	}, nil
+}
+
+func (s *Service) getInstanceByToken(ctx context.Context) (*fastapi.Instance, error) {
+	token := metautils.ExtractIncoming(ctx).Get("X-fpa-Token")
```

Comment:
Maybe I just missed this, or maybe it's still to be seen below, but how does a client acquire an auth token? Is it only when `Create` is called you get an auth token back? Is there a way to get a new auth token if necessary?